fractal_server/app/routes/api/v2/workflow_import.py

```diff
@@ -0,0 +1,230 @@
+from ....db import get_async_db
+from ....models.v2 import TaskV2
+from ....models.v2 import WorkflowV2
+from ....schemas.v2 import TaskImportV2Legacy
+from ....schemas.v2 import WorkflowImportV2
+from ....schemas.v2 import WorkflowReadV2
+from ....schemas.v2 import WorkflowTaskCreateV2
+from ._aux_functions import _check_workflow_exists
+from ._aux_functions import _get_project_check_owner
+from ._aux_functions import _workflow_insert_task
+from fractal_server.app.models import LinkUserGroup
+from fractal_server.app.models import UserOAuth
+from fractal_server.app.models.security import UserGroup
+from fractal_server.app.models.v2.task import TaskGroupV2
+from fractal_server.app.routes.auth import current_active_user
+from fractal_server.app.security import FRACTAL_DEFAULT_GROUP_NAME
+
+router = APIRouter()
+
+
+async def _get_user_tasks(user_id: int, db: AsyncSession) -> list[TaskV2]:
+    """Retrieve tasks that belong to the user."""
+    stm = (
+        select(TaskV2)
+        .join(TaskGroupV2)
+        .where(TaskGroupV2.id == TaskV2.taskgroupv2_id)
+        .where(TaskGroupV2.user_id == user_id)
+    )
+    res = await db.execute(stm)
+    task_user_list = res.scalars().all()
+    return task_user_list
+
+
+async def _get_default_group_id(db: AsyncSession) -> int:
+    """Get the default user group ID."""
+    stm = select(UserGroup).where(UserGroup.name == FRACTAL_DEFAULT_GROUP_NAME)
+    res = await db.execute(stm)
+    default_group_id = res.scalars().one().id
+    return default_group_id
+
+
+async def _find_task_by_source_or_version(
+    wf_task, user, db, task_user_list, default_group_id
+) -> Optional[int]:
+    """Find a task by source or version."""
+    if isinstance(wf_task.task, TaskImportV2Legacy):
+        task = await _get_task_by_source(wf_task, user, db)
+    else:
+        task = await _get_task_by_version(
+            wf_task, user, db, task_user_list, default_group_id
+        )
+    if not task:
+        raise HTTPException(
+            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+            detail="Expected one TaskV2 to match the imported one.",
+        )
+
+    return task.id
+
+
+async def _get_task_by_source(wf_task, user, db) -> Optional[TaskV2]:
+    """Find a task by its source."""
+    if wf_task.task.source is None:
+        return None
+
+    stm_source = (
+        select(TaskV2)
+        .join(TaskGroupV2)
+        .where(TaskGroupV2.id == TaskV2.taskgroupv2_id)
+        .where(TaskV2.source == wf_task.task.source)
+        .where(
+            or_(
+                TaskGroupV2.user_id == user.id,
+                TaskGroupV2.user_group_id.in_(
+                    select(LinkUserGroup.group_id).where(
+                        LinkUserGroup.user_id == user.id
+                    )
+                ),
+            )
+        )
+    )
+    res = await db.execute(stm_source)
+    task = res.scalars().one_or_none()
+    return task
+
+
+async def _get_task_by_version(
```

Comment:
(all working on a Python list)
1. Construct a sublist of tasks with matching `pkg_name` and `name`
2. If version is unset, find the latest from the filtered list
3. Now that version is known, construct a sublist of tasks with matching version
4. Disambiguate by ownership, if needed